### PR TITLE
add xkbcommon/0.10.0

### PR DESCRIPTION
**Pull request by ericLemanissier**
Specify library name and version:  **xkbcommon/0.10.0**

this is a port of https://github.com/bincrafters/conan-xkbcommon

- [X] I've read the [guidelines](https://github.com/conan-io/conan-center-index/wiki#how-to-submit-a-pull-request) for contributing.
- [X] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [X] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [X] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by ericLemanissier** [comment]
I can see 3 different solutions for xkbcommon:

1. use a regular package, like this PR does currently
2. make it a xkbcommon/system package which just installs the system package
3. include xkbcommon in xorg/system

I'm not sure which solution is the best.

**Comment by SSE4** [comment]
as I understand, it may use either X11 or wayland as backend:
```
To build for use with Wayland, you can disable X11 support while still using the X11 keyboard configuration resource files thusly:
```
in such case, I don't think it should be part of xorg package.

**Comment by Croydon** [comment]
I see that you change it to a system package. So did it become a clear case that this is the best move @ericLemanissier?

**Comment by ericLemanissier** [comment]
my current understanding is that xkbcommon has a "core" library which does not require xorg libraries. There is also the libxkbcommon-x11 library which depends on xcb. None of these use wayland as a backend, but wayland does require the "core" xkbcommon lib (explained [here](https://github.com/xkbcommon/libxkbcommon/blob/afb26e7df9090a0b765eb294b6efff448f763b6f/PACKAGING#L66)). The "enable-wayland" option is actually building [a demo program](https://github.com/xkbcommon/libxkbcommon/blob/2cb90c958f7efa36b00fe845c38bbe0d00e15b1e/meson.build#L540)

In the end I think it's not really important if xkbcommon is a system or a usual package, but I'd say it should have a `with_x11` option (True by default) which enables installation/compilation of libxkbcommon-x11. wayland consumers can set xkbcommon:with_x11=False to remove the xorg dependancy

**Comment by ericLemanissier** [comment]
I'd like to moderate myself regarding the with_x11 option above. It would probably be an unneeded complication, just to save on system package installation.
@SSE4 @uilianries what's your take on this one ?

**Comment by SSE4** [comment]
@ericLemanissier I'd follow KISS principle for the first implementation until everyone actually requests an option

**Comment by ericLemanissier** [comment]
Ok, that's the current state of the PR

File: recipes/xkbcommon/all/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+                packages = ["libxkbcommon-dev", "libxkbcommon-x11-dev"]
+            elif tools.os_info.with_yum or tools.os_info.with_dnf:
+                packages = ["libxkbcommon-devel", "libxkbcommon-x11-devel"]
+            elif tools.os_info.with_pacman:
+                packages = ["libxkbcommon", "libxkbcommon-x11"]
+            elif tools.os_info.with_zypper:
+                packages = ["libxkbcommon-devel", "libxkbcommon-x11-devel"]
+            else:
+                self.output.warn("Do not know how to install 'xkbcommon' for {}.".format(tools.os_info.linux_distro))
+            for p in packages:
```
Comment (SSE4) [inline recipes/xkbcommon/all/conanfile.py:51]:
```suggestion
            for p in packages if packages else []:
```

Comment (ericLemanissier) [reply recipes/xkbcommon/all/conanfile.py:51]: :D how did I forget that

File: recipes/xkbcommon/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+            elif tools.os_info.with_pacman:
+                packages = ["libxkbcommon", "libxkbcommon-x11"]
+            elif tools.os_info.with_zypper:
+                packages = ["libxkbcommon-devel", "libxkbcommon-x11-devel"]
+            else:
+                self.output.warn("Do not know how to install 'xkbcommon' for {}.".format(tools.os_info.linux_distro))
+            for p in packages if packages else []:
+                package_tool.install(update=True, packages=p)
+
+    def package_info(self):
```
Comment (SSE4) [inline recipes/xkbcommon/all/conanfile.py:54]:
```suggestion
    def package_info(self):
        self.cpp_info.includedirs = []
        self.cpp_info.libdirs = []
```

File: recipes/xkbcommon/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+
+class XkbcommonConan(ConanFile):
+    name = "xkbcommon"
+    description = "keymap handling library for toolkits and window systems"
+    topics = ("conan", "xkbcommon", "keyboard")
+    url = "https://github.com/conan-io/conan-center-index"
+    homepage = "https://github.com/xkbcommon/libxkbcommon"
+    license = "MIT"
+    settings = {"os": "Linux"}
```
Comment (uilianries) [inline recipes/xkbcommon/all/conanfile.py:12]: I prefer moving it to configure. It's not the first time we have a similar case. The advantage there is throwing a ConanInvalidConfiguration and write a specific and detailed message.

**Comment by Croydon** [comment]
## Stop⚠️

As far as I see it we are still not sure if this should be a system package. 
Packages should be only system packages when it is a clear case to us.
System packages should stay an huge exception to regular packages

I don't think that "it exist only for Linux" is an argument for system packages either. 
The question is if xkbcommon is a low-level package, which would conflict with existing system installations of xkbcommon 

Unfortunately, I don't have an answer for that question myself as I'm not familiar with xkbcommon, but please, when in doubt it should not be a system package

**Comment by ericLemanissier** [comment]
I asked upstream: https://github.com/xkbcommon/libxkbcommon/issues/166 and there does not seem to be strong argument against making xkbcommon a regular common package.
I can see two arguments in favor of a regular package:

1. https://github.com/bincrafters/conan-xkbcommon has been a regular package from the start, and we never had reports of conflict with system xkbcommon
2. qt has had bundled sources for xkbcommon for years https://codereview.qt-project.org/c/qt/qtbase/+/53496 - https://codereview.qt-project.org/c/qt/qtbase/+/245836, and the only reason they removed it is because it removes maintenance burden (and because all the platforms supported by qt have system xkbcommon)

**Comment by ericLemanissier** [comment]
this is blocked by https://github.com/conan-io/conan-center-index/pull/2255

**Comment by Croydon** [comment]
@ericLemanissier Thanks for your continuing work on this!

**Comment by danimtb** [comment]
Merged #2255!

**Comment by ericLemanissier** [comment]
retrigger CI

File: recipes/xkbcommon/all/test_package/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+    generators = "cmake"
+
+    def build(self):
+        cmake = CMake(self)
+        cmake.configure()
+        cmake.build()
+
+    def test(self):
+        bin_path = os.path.join("bin", "test_package")
+        self.run(bin_path, run_environment=True)
```
Comment (uilianries) [inline recipes/xkbcommon/all/test_package/conanfile.py:16]:
```suggestion
        if not tools.cross_building(self.settings):
            bin_path = os.path.join("bin", "test_package")
            self.run(bin_path, run_environment=True)
```

File: recipes/xkbcommon/all/test_package/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,17 @@
+from conans import ConanFile, CMake
```
Comment (uilianries) [inline recipes/xkbcommon/all/test_package/conanfile.py:1]:
```suggestion
from conans import ConanFile, CMake, tools
```

**Comment by SSE4** [comment]
seems like we don't have a strong argument yet for making it system vs regular package... reading https://github.com/xkbcommon/libxkbcommon/issues/166 I am still not completely sure.

e.g. on Windows, there are low-level hardware-specific libraries for input sources management, most famous are `Xinput` and `DirectInput` - they provide API to access keyboard, joystick, gamepad, wheel and other input sources.
for me this falls into the same category as other low-level APIs, e.g. to access GPU resources: 
- rendering (OpenGL, Vulkan, Direct3D, DirectDraw)
- computing (DirectCompute, OpenCL, CUDA)
- video processing (MFX/QSV, VAAPI, VDPAU, DXVA)

same probably for low-level audio APIs (such as alsa, jack, pulse, nas, esd, oss, etc.).

we also use system packages for X11/Xorg libraires. while they are completely user-space and relies on lower-lever graphics abstraction layers (dri, drm, mesa libraries), it's still too much burden and too many issues with libraries built from conan in our way.

for xkb-common - yes, it might sound like it's direct equivalent of DirectInput/XInput for linux, however, as stated by developers, it doesn't directly communicate with vendor-specific hardware. 

so for this one I am not absolutely sure right now.

**Comment by danimtb** [comment]
I see there are no strong reasons to go for one approach or the other. In that case, I'd like to have it as a normal Conan package, so you have full control over the build process and we make sure it complies with the checks of a regular package as well. WDYT?

**Pull request merged**